File: sdk/canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/ledger/api/messages/command/submission/SubmitReassignmentRequest.scala
// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.digitalasset.canton.ledger.api.messages.command.submission

import com.digitalasset.canton.topology.SynchronizerId
import com.digitalasset.canton.util.ReassignmentTag.{Source, Target}
import com.digitalasset.daml.lf.data.{Ref, Time}
import com.digitalasset.daml.lf.value.Value

final case class SubmitReassignmentRequest(
    submitter: Ref.Party,
    applicationId: Ref.ApplicationId,
    commandId: Ref.CommandId,
    submissionId: Ref.SubmissionId,
    workflowId: Option[Ref.WorkflowId],
    reassignmentCommand: Either[AssignCommand, UnassignCommand],
)

final case class UnassignCommand(
    sourceSynchronizerId: Source[SynchronizerId],
    targetSynchronizerId: Target[SynchronizerId],
    contractId: Value.ContractId,
)
final case class AssignCommand(
    sourceSynchronizerId: Source[SynchronizerId],
    targetSynchronizerId: Target[SynchronizerId],
    unassignId: Time.Timestamp,
)

File: sdk/canton/community/bindings-java/src/main/java/com/daml/ledger/javaapi/data/SubmitReassignmentRequest.java
// Copyright (c) 2025 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.daml.ledger.javaapi.data;

import com.daml.ledger.api.v2.CommandSubmissionServiceOuterClass;
import org.checkerframework.checker.nullness.qual.NonNull;

public final class SubmitReassignmentRequest {

  public static CommandSubmissionServiceOuterClass.SubmitReassignmentRequest toProto(
      @NonNull ReassignmentCommands command) {
    return CommandSubmissionServiceOuterClass.SubmitReassignmentRequest.newBuilder()
        .setReassignmentCommands(command.toProto())
        .build();
  }

  public static ReassignmentCommands fromProto(
      CommandSubmissionServiceOuterClass.@NonNull SubmitReassignmentRequest request) {
    return ReassignmentCommands.fromProto(request.getReassignmentCommands());
  }
}
